src/Engine/EmitMSIL/Replication.cs

```diff
@@ -315,7 +351,114 @@ private static List<List<ReplicationGuide>> ConstructRepGuides(string[][] replic
             return compliantTarget;
         }
 
+        private static CLRFunctionEndPoint SelectFEPFromMultiple(
+            CLRStackValue thisPtr,
+            MSILRuntimeCore runtimeCore,
+            List<CLRFunctionEndPoint> feps,
+            List<CLRStackValue> argumentsList)
+        {
+            Validity.Assert(thisPtr.IsNull || thisPtr.IsPointer,
+                            "this pointer wasn't null or a pointer. {89635B06-AD53-4170-ADA5-065EB2AE5858}");
+
+            var staticFuncClassId = feps.FirstOrDefault().procedureNode.ClassID;
+
+            // We have multiple possible scopes for the function call:
+            // 1. Static method call - no this pointer
+            // ex: ClassA.Method();
+            //    Hidden static methods generate multiple feps.
+            //    We do not need to check actually if the method has the "IsHideBySig" (https://docs.microsoft.com/en-us/dotnet/api/system.reflection.methodbase.ishidebysig)
+            //    because static methods can only be hidden.
+            // 2. Method call from an instance of a class - valid this pointer.
+            // ex: classAInstance.method();
+            // 3. Function from the global scope - no this pointer and no class scope.
+            // ex: SomeGlobalFunction();
+            //
+            // All 3 cases will run through the same matching steps.
+
+            // A static function call has an invalid this pointer and a valid class scope;
+            bool isValidStaticFuncCall = !thisPtr.IsNull && staticFuncClassId != Constants.kInvalidIndex;
+            if (isValidStaticFuncCall)
+            {
+                return feps.FirstOrDefault();
+            }
+
+            int typeID = thisPtr.TypeUID;
+
+            // Try to match with feps belonging to the class scope (most derived class should have priority).
+            // In this case we simply select the function that belongs to the calling class.
+            // The assumption here is that all function end points in "feps" have already been checked that they have the same signature.
+            IEnumerable<CLRFunctionEndPoint> exactFeps = feps.Where(x => x.procedureNode.ClassID == typeID);
+            if (exactFeps.Count() == 1)
+            {
+                return exactFeps.First();
+            }
+
+            //Walk the class tree structure to find the method
+            while (runtimeCore.ClassTable.ClassNodes[typeID].Base != Constants.kInvalidIndex)
+            {
+                typeID = runtimeCore.ClassTable.ClassNodes[typeID].Base;
+
+                foreach (var fep in feps)
```

Comment:
can you add `{`

Reply:
how does this logic factor into this comment:
`(most derived class should have priority).`

Reply:
this entire function has been copy pasted (and retrofitted) to the new VM.
For the most derived class priority :
I think the initial check  `feps.Where(x => x.procedureNode.ClassID == typeID);` represents "the most derived class".
If no exact match is found, then we go through the base classes